### ISSUE TITLE
Different accuracy when running distributed training

### ISSUE THREAD

**Issue by cotterpl** [microsoft/LightGBM#3473]
It is not a bug but rather a question. I am testing LightGBM behaviour in a few scenarios:

1. Local: Train locally
2. Distributed: Train in distributed manner on two Kubernetes pods that run on Dask cluster
3. One worker: Train on the same Kubernetes pods with Dask cluster but use only 1 worker (no distribution)

When running distributed training I am using tree_learner=data. In the distributed scenario this data is partitioned between dask workers.

I have noticed that when LightGBM is trained in a distributed manner it behaves differently. I have made 5 runs for each of the tested scenarios:

The mean absolute error on test set is significantly worse for the distributed scenario:
- Local MAE: ~777
- Distributed MAE: ~792
- One worker MAE: ~779

If I use early stopping it builds less trees:
- Local: ~3800 trees, 
- Distributed: ~600 trees

I have tried not using early stopping and reducing learning rate but nothing seems to help. The issue seems to lie in the nature of distributed training. The distributed training is always significantly worse than local one. I was expecting similar scores. Is that expected behaviour?

**Comment by guolinke** [comment]
@cotterpl did you use the same validation set (without partition) in different nodes?

**Comment by cotterpl** [comment]
Yes, test set is exactly the same. Training set is the same (but partitioned for the distributed case). When using early stopping evaluation set is also the same - replicated in each worker. All hyperparameters are also the same.

Obtained MAE scores and amounts of trees are means from 5 runs and the score difference can not be explained by its standard deviation.

**Comment by guolinke** [comment]
can you also provide the number of data samples? 

In distributed mode, the feature bucket (binning) is also distributed. Assuming there are `m` features and `m` machines, each machine will only bucket one feature, by its local partitioned data.

In single-node mode, all features are buckted by non-partition data.

I think the above difference, cause the different feature binning results. And the feature binning in single-node is better, thus its accuracy is also better.

Therefore, in distributed mode, it is better to ensure each node have enough data, for feature binning.

**Comment by cotterpl** [comment]
I am running it in 2 nodes setup. Data is provided by DASK cluster.

The overall data size:
385918 rows x 19 columns

I have checked the data that enters lightgbm training in each of the nodes. The data has all of the features on both of them. For some reason dask does not distribute the rows evenly between nodes. Node 1 has been provided with 332098 while node 2 has been provided with only 53820 rows of data.

The hyperparameters that I am using:

```
objective = 'mae'
    return {
        'n_estimators': 4000,
        'task': 'train',
        'boosting_type': 'gbdt',
        'objective': objective,
        'metric': objective,
        'max_depth': 8,
        'min_child_samples': 30,
        'num_leaves': 25,
        'learning_rate': 0.02,
        'colsample_bytree': 0.7,
        'subsample': 0.5,
        'subsample_freq': 1,
        'bagging_seed': seed,
        'verbose': -1,
        'seed': seed,
        'silent': True,
        'importance_type': 'gain',
        'tree_learner': 'data',
    }
```

**Comment by guolinke** [comment]
@cotterpl this un-even partition will cause half of the features don't have accuracy feature binning, and cause bad performance. 
besides, 385918 is relatively small, and I don't recommend to use it in distributed mode.

**Comment by cotterpl** [comment]
I have figured out that we are not shuffling the set properly. Doing shuffle fixed the issue. Now score is similar to locally run one.
